rfcs/react-components/convergence/slot-render-as.md

```diff
@@ -0,0 +1,183 @@
+# RFC: Use `as` instead of `children` for slot render function
+
+Contributors: behowell
+
+## Summary
+
+This proposes modifying the slot render function API to pass a render function via the slot's `as` prop instead of the `children` prop. This has several benefits, the primary of which is that it allows the user to write a render function for a slot without overriding the default children of a slot.
+
+This is a complementary proposal to bsunderhus's [RFC #27164: slot children render function support](https://github.com/microsoft/fluentui/pull/27164). It provides an alternative way to avoid having a render function override a slot's default `children` prop. This RFC has no dependencies on it, and could be adopted separately if needed.
+
+## Background
+
+Slots accept render functions that allow them to override the component and/or props used when rendering the slot. For more background, see https://react.fluentui.dev/?path=/docs/concepts-developer-customizing-components-with-slots--page#replacing-the-entire-slot
+
+The render function is passed in as the slot's `children` prop.
+
+## Problem statement
+
+This RFC seeks to solve some problems with render functions as they are today.
+
+> ℹ️ NOTE: [RFC #27164: slot children render function support](https://github.com/microsoft/fluentui/pull/27164) solves _additional_ problems with slot render functions beyond what's listed here. The two RFCs complement each other; but don't solve all of the same issues.
+
+1. Using `children` as the render function will override any default children provided by the component's internals. (This is listed as Problem 1 in [RFC #27164](https://github.com/microsoft/fluentui/pull/27164)).
+2. The root slot's render function conflicts with the use of children as a render-props function.
+
+   - For example, `Field`'s children can be a render function that is given props to spread on the child:
+     ```jsx
+     // Field uses a custom render-props style function as its children
+     <Field>{props => <Input {...props} />}</Field>
+     ```
+   - However, that now makes it impossible to add a render function that _replaces_ the root slot, which would typically be done as:
+     ```jsx
+     // Label doesn't override the root slot's render function
+     <Label>{(Component, props) => <MyLabelComponent {...props} />}</Label>
```

Comment:
- This works because there is no special handling for `root` slot
- This does not throw a TS error because `ReactNode` definition that we use is [a black hole](https://changelog.com/posts/the-react-reactnode-type-is-a-black-hole) before `@types/react@18`

```tsx
// @types/react@17 PASSING 🙈
const props: { children: React.ReactNode } = {
  children: () => null,
}
```

```tsx
// @types/react@18 THROWS 👍 
const props: { children: React.ReactNode } = {
  children: () => null,
}
```

⬆️ https://codesandbox.io/s/cocky-architecture-osickm?file=/src/App.tsx

I don't think that this usage was ever advertised as something recommended, correct me if I am wrong.

Reply:
I don't know whether or not it was intended to support a render function for the root slot. But given that it was technically supported before, I don't see any reason to remove support? With this RFC, I don't think there would be any type issues like you described in your comment.

Is your comment a reason to change or reject the RFC? If so, could you elaborate?